spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -527,9 +425,8 @@ object SparkTableUtil {
     val metricsConfig = MetricsConfig.fromProperties(targetTable.properties)
 
     val manifests = partitionDS
-      .flatMap(partition => listPartition(partition, serializableConf, metricsConfig))
+      .flatMap(partition => listPartition(partition, spec, serializableConf, metricsConfig))
       .repartition(numShufflePartitions)
-      .orderBy($"path")
```

Comment:
Why do we want to remove the sort? I think it is needed to collocate files for the same partition next to each other so that partition skipping is quick.

Reply:
The `GenericDataFile` is not a case class so it doesn't have a schema with its fields. It just has one column named `value`. I tried to construct an extra column with UDF like:

```scala
    val toPath = org.apache.spark.sql.functions.udf {s:DataFile => s.path().toString}

    val manifests = partitionDS
      .flatMap(partition => listPartition(partition, spec, serializableConf, metricsConfig))
      .repartition(numShufflePartitions)
      .orderBy(toPath($"value"))
      .mapPartitions(buildManifest(serializableConf, spec, stagingDir))
      .collect()

```
but it throws 
`java.lang.ClassCastException: [B cannot be cast to org.apache.iceberg.DataFile`. So I remove this, do you have any idea on this?

Reply:
BTW, even we coalesce files in a manifest for same partition, I think we still have to iterate through all manifest entries in the manifest for partition skipping.  Please correct me if I am wrong.

Reply:
What about this? It is a bit uglier than what we had before but should work.

```
implicit val manifestFileEncoder: Encoder[ManifestFile] = Encoders.javaSerialization[ManifestFile]
implicit val dataFileEncoder: Encoder[DataFile] = Encoders.javaSerialization[DataFile]
implicit val pathDataFileEncoder: Encoder[(String, DataFile)] = Encoders.tuple(Encoders.STRING, dataFileEncoder)

....

val manifests = partitionDS
  .flatMap(partition => listPartition(partition, spec, serializableConf, metricsConfig))
  .repartition(numShufflePartitions)
  .map(file => (file.path.toString, file))
  .orderBy($"_1")
  .mapPartitions(files => buildManifest(serializableConf, spec, stagingDir)(files.map(_._2)))
  .collect()
```

The reason sort is important is because we want to filter out irrelevant manifests when planning a job with a partition predicate. Each manifest keeps lower/upper bounds for partition values. If we place all files for the same partition in one manifest, we will need to read only one manifest to plan a job for that partition. If we don't sort files before creating manifests, we risk having files for one partition scattered across many manifests.

Reply:
Nice, I tried your way but forgot defined the tuple encoder.